BUG: optimize.minimize: `trust-constr` minimization not enforcing constraint

### Describe your issue.

I am performing a `trust-constr` minimization with a `LinearConstraint` where `keep_feasible = True`. My expectation is that the variables passed into the objective function should never violate the constraint. However, I ran into a case where the variables did violate the constraint.

I have provided a reproducing code example below. To keep things simple, I wrote the code example to just return the offending objective function and Jacobian values, which I pulled from my actual code run. As you can see, the assertion fails because an infeasible set of variables was passed into the objective function.

### Reproducing Code Example

```python
import numpy
from numpy.typing import NDArray
from scipy.optimize import minimize, Bounds, LinearConstraint


_OBJECTIVE_AND_JACOBIAN_VALUES = [
    (-1577.7743237769168, (64.09380887783628, 58.57113040751609, 131.4168010103574)),
    (-1613.217804207217, (2.1886161649781815, 18.817148203999555, 3.378215339854541)),
    (-1613.9641022007631, (-1.9901246302386326, 9.178839374990384, -8.302130929917746)),
    (-1614.7670317747486, (-4.8002109006041005, -4.098931430224862, -15.33447306093558)),
    (-1615.0973693437152, (-1.6461760827423748, -3.696351982768015, -8.343229694079792)),
    (-1615.194458754922, (1.108317234151853, 0.21742493238622587, -2.3727009599767945)),
    (-1615.300132870974, (2.368343175056354, 1.5511270059237487, 0.4585017357582592)),
    (-1615.4615740556708, (9.464282738395625, 18.306759981137354, 23.92768915405854)),
    (-1615.7359999237315, (3.5145051607575764, -0.6229557128333122, 3.648824522974539)),
    (-1615.966608643255, (2.804918769311262, 1.0721176544913282, 2.342082157657089)),
    (-1615.2517574460453, (12.405099581907178, 65.3065228394017, 78.11332333252767)),
    (-1616.2899332077368, (-1.013885569127406, -7.797014581828413, -10.424283781969185)),
    (-1616.4561204619968, (-0.47762150012714866, -6.237376229258435, -8.779865924960948)),
    (-1616.6383117710238, (0.08073154387888869, -4.333899571985981, -6.698728068630178)),
    (-1616.8247415154556, (0.8001222456113003, -1.3050249397432074, -3.2695792410868867))
]


def _main():
    iteration_count = 0

    def objective_and_jacobian(
            theta: NDArray[float]
    ) -> tuple[float, NDArray[float]]:
        nonlocal iteration_count

        constraint_value = theta[1] + theta[2]
        assert constraint_value < 1, \
            f"Constraint violated after {iteration_count} iterations with value {constraint_value}."

        objective, jacobian = _OBJECTIVE_AND_JACOBIAN_VALUES[iteration_count]
        iteration_count += 1

        return objective, numpy.array(jacobian)

    minimize(
        fun=objective_and_jacobian,
        x0=[
            -10.743527050727764,
            0.45,
            0.3
        ],
        method='trust-constr',
        jac=True,
        bounds=Bounds(
            lb=[-numpy.inf, 0.0, 0.0],
            ub=[numpy.inf, 1.0, 1.0],
            keep_feasible=True
        ),
        constraints=LinearConstraint(
            A=[0, 1, 1],
            ub=0.999999,
            keep_feasible=True
        )
    )


if __name__ == '__main__':
    _main()
```

### Error message

```shell
Traceback (most recent call last):
  File "/Users/fumoboy007/Desktop/KeepFeasibleBugRepro/main.py", line 65, in <module>
    _main()
  File "/Users/fumoboy007/Desktop/KeepFeasibleBugRepro/main.py", line 42, in _main
    minimize(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_minimize.py", line 799, in minimize
    res = _minimize_trustregion_constr(fun, x0, args, jac, hess, hessp,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py", line 546, in _minimize_trustregion_constr
    _, result = tr_interior_point(
                ^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_trustregion_constr/tr_interior_point.py", line 336, in tr_interior_point
    z, state = equality_constrained_sqp(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_trustregion_constr/equality_constrained_sqp.py", line 161, in equality_constrained_sqp
    f_next, b_next = fun_and_constr(x_next)
                     ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_trustregion_constr/tr_interior_point.py", line 95, in function_and_constraints
    f = self.fun(x)
        ^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_differentiable_functions.py", line 393, in fun
    self._update_x(x)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_differentiable_functions.py", line 350, in _update_x
    self._update_hess()
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_differentiable_functions.py", line 384, in _update_hess
    self._update_grad()
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_differentiable_functions.py", line 375, in _update_grad
    self.g = self._wrapped_grad(self.x, f0=self.f)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_differentiable_functions.py", line 39, in __call__
    g = np.atleast_1d(self.grad(np.copy(x), *self.args))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_optimize.py", line 84, in derivative
    self._compute_if_needed(x, *args)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/python3.11/site-packages/scipy/optimize/_optimize.py", line 74, in _compute_if_needed
    fg = self.fun(x, *args)
         ^^^^^^^^^^^^^^^^^^
  File "/Users/fumoboy007/Desktop/KeepFeasibleBugRepro/main.py", line 34, in objective_and_jacobian
    assert constraint_value < 1, \
           ^^^^^^^^^^^^^^^^^^^^
AssertionError: Constraint violated after 15 iterations with value 1.011765851400075.
```

### SciPy/NumPy/Python version and system information

```shell
1.17.0 2.3.5 sys.version_info(major=3, minor=11, micro=14, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    has ilp64: false
    include directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include
    lib directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib
    name: blas
    openblas configuration: unknown
    pc file directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/pkgconfig
    version: 3.9.0
  lapack:
    detection method: pkgconfig
    found: true
    has ilp64: false
    include directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include
    lib directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib
    name: lapack
    openblas configuration: unknown
    pc file directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib/pkgconfig
    version: 3.9.0
  pybind11:
    detection method: pkgconfig
    include directory: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include
    name: pybind11
    version: 3.0.1
Compilers:
  c:
    args: -ftree-vectorize, -fPIC, -fstack-protector-strong, -O2, -pipe, -isystem,
      /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/work=/usr/local/src/conda/scipy-split-1.17.0,
      -fdebug-prefix-map=/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro=/usr/local/src/conda-prefix,
      -D_FORTIFY_SOURCE=2, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -mmacosx-version-min=11.0, -mmacosx-version-min=11.0
    commands: arm64-apple-darwin20.0.0-clang
    linker: ld64
    linker args: -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs, -Wl,-rpath,/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib,
      -L/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib, -ftree-vectorize,
      -fPIC, -fstack-protector-strong, -O2, -pipe, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/work=/usr/local/src/conda/scipy-split-1.17.0,
      -fdebug-prefix-map=/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro=/usr/local/src/conda-prefix,
      -D_FORTIFY_SOURCE=2, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -mmacosx-version-min=11.0, -mmacosx-version-min=11.0
    name: clang
    version: 19.1.7
  c++:
    args: -ftree-vectorize, -fPIC, -fstack-protector-strong, -O2, -pipe, -stdlib=libc++,
      -fvisibility-inlines-hidden, -fmessage-length=0, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/work=/usr/local/src/conda/scipy-split-1.17.0,
      -fdebug-prefix-map=/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro=/usr/local/src/conda-prefix,
      -D_FORTIFY_SOURCE=2, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -mmacosx-version-min=11.0, -mmacosx-version-min=11.0
    commands: arm64-apple-darwin20.0.0-clang++
    linker: ld64
    linker args: -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs, -Wl,-rpath,/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib,
      -L/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib, -ftree-vectorize,
      -fPIC, -fstack-protector-strong, -O2, -pipe, -stdlib=libc++, -fvisibility-inlines-hidden,
      -fmessage-length=0, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/work=/usr/local/src/conda/scipy-split-1.17.0,
      -fdebug-prefix-map=/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro=/usr/local/src/conda-prefix,
      -D_FORTIFY_SOURCE=2, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -mmacosx-version-min=11.0, -mmacosx-version-min=11.0
    name: clang
    version: 19.1.7
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.2.4
  fortran:
    args: -march=armv8.3-a, -ftree-vectorize, -fPIC, -fno-stack-protector, -O2, -pipe,
      -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/work=/usr/local/src/conda/scipy-split-1.17.0,
      -fdebug-prefix-map=/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro=/usr/local/src/conda-prefix
    commands: /Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/_build_env/bin/arm64-apple-darwin20.0.0-gfortran
    linker: ld64
    linker args: -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs, -Wl,-rpath,/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib,
      -L/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/lib, -march=armv8.3-a,
      -ftree-vectorize, -fPIC, -fno-stack-protector, -O2, -pipe, -isystem, /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/include,
      -fdebug-prefix-map=/Users/runner/miniforge3/conda-bld/scipy-split_1768134621190/work=/usr/local/src/conda/scipy-split-1.17.0,
      -fdebug-prefix-map=/opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro=/usr/local/src/conda-prefix
    name: gcc
    version: 14.3.0
  pythran:
    include directory: ../../_build_env/venv/lib/python3.11/site-packages/pythran
    version: 0.18.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: darwin
  cross-compiled: true
  host:
    cpu: arm64
    endian: little
    family: aarch64
    system: darwin
Python Information:
  path: /opt/homebrew/Caskroom/miniforge/base/envs/keep-feasible-bug-repro/bin/python
  version: '3.11'
```

--- a/scipy/optimize/_constraints.py
+++ b/scipy/optimize/_constraints.py
@@ -71,9 +71,11 @@ class NonlinearConstraint:
         ``hess(x, v) -> {LinearOperator, sparse array, array_like}, shape (n, n)``.
         Here ``v`` is ndarray with shape (m,) containing Lagrange multipliers.
     keep_feasible : array_like of bool, optional
-        Whether to keep the constraint components feasible throughout
-        iterations. A single value sets this property for all components.
-        Default is False. Has no effect for equality constraints.
+        Whether to keep the constraint components feasible at the end of each
+        iteration. A single value sets this property for all components.
+        Default is False. Has no effect for equality constraints. Note that intervening
+        computations of the objective (not constraint) function may be located within
+        an infeasible region
     finite_diff_rel_step : None or array_like, optional
         Relative step size for the finite difference approximation. Default is
         None, which will select a reasonable value automatically depending
@@ -153,9 +155,11 @@ class LinearConstraint:
         `lb` and `ub` as  necessary. Defaults to ``lb = -np.inf``
         and ``ub = np.inf`` (no limits).
     keep_feasible : dense array_like of bool, optional
-        Whether to keep the constraint components feasible throughout
-        iterations. A single value sets this property for all components.
-        Default is False. Has no effect for equality constraints.
+        Whether to keep the constraint components feasible at the end of each
+        iteration. A single value sets this property for all components.
+        Default is False. Has no effect for equality constraints. Note that intervening
+        computations of the objective (not constraint) function may be located within
+        an infeasible region
     """
     def _input_validation(self):
         if self.A.ndim != 2:

--- a/scipy/optimize/_constraints.py
+++ b/scipy/optimize/_constraints.py
@@ -75,7 +75,7 @@ class NonlinearConstraint:
         iteration. A single value sets this property for all components.
         Default is False. Has no effect for equality constraints. Note that intervening
         computations of the objective (not constraint) function may be located within
-        an infeasible region
+        an infeasible region.
     finite_diff_rel_step : None or array_like, optional
         Relative step size for the finite difference approximation. Default is
         None, which will select a reasonable value automatically depending
@@ -159,7 +159,7 @@ class LinearConstraint:
         iteration. A single value sets this property for all components.
         Default is False. Has no effect for equality constraints. Note that intervening
         computations of the objective (not constraint) function may be located within
-        an infeasible region
+        an infeasible region.
     """
     def _input_validation(self):
         if self.A.ndim != 2:

